src/cpp/README.md

```diff
@@ -30,20 +25,75 @@ To add gRPC as a dependency in bazel:
   grpc_deps()
   ```
 
-## cmake
+## CMake
+
+`cmake` is your best option if you cannot use bazel. It supports building on Linux,
+MacOS and Windows (official support) but also has a good chance of working on
+other platforms (no promises!). `cmake` has good support for crosscompiling and
+can be used for targeting the Android platform.
 
-`cmake` is your best option if you cannot use bazel. It supports building on Linux, MacOS and Windows (official support) but also has a good chance of working on other platforms (no promises!). `cmake` has good
-support for crosscompiling and can be used for targeting Android platform.
+To build gRPC C++ from source, follow the [BUILDING guide](../../BUILDING.md).
 
-If your project is using cmake, there are several ways to add gRPC dependency.
-- install gRPC via cmake first and then locate it with `find_package(gRPC CONFIG)`. [Example](../../examples/cpp/helloworld/CMakeLists.txt)
-- via cmake's `ExternalProject_Add` using a technique called "superbuild". [Example](../../examples/cpp/helloworld/cmake_externalproject/CMakeLists.txt)
-- add gRPC source tree to your project (preferably as a git submodule) and add it to your CMake project with `add_subdirectory`. [Example](../../examples/cpp/helloworld/CMakeLists.txt)
+### find_package
 
-If your project is not using CMake (e.g. you're using `make` directly), you can first install gRPC C++ using CMake,
-and have your non-CMake project rely on the `pkgconfig` files which are provided by gRPC installation. [Example](../../test/distrib/cpp/run_distrib_test_cmake_pkgconfig.sh)
+The canonical way to discover dependencies in CMake is the
+[`find_package` command](https://cmake.org/cmake/help/latest/command/find_package.html).
+
+```cmake
+find_package(gRPC CONFIG REQUIRED)
+add_executable(my_exe my_exe.cc)
+target_link_libraries(my_exe gRPC::grpc++)
+```
+[Full example](../../examples/cpp/helloworld/CMakeLists.txt)
+
+`find_package` can only find software that has already been installed on your
+system. The following sections describe strategies to automatically build gRPC
+as part of your project.
+
+### FetchContent
+If you are using CMake v3.11 or newer you should use CMake's
+[FetchContent module](https://cmake.org/cmake/help/latest/module/FetchContent.html).
+This will download gRPC's source code when you initially configure your project.
+
+```cmake
+include(FetchContent)
+FetchContent_Declare(
+  gRPC
+  GIT_REPOSITORY https://github.com/grpc/grpc
+)
+FetchContent_MakeAvailable(gRPC)
```

Comment:
I am not very familiar with FetchContent, so I have a few questions:
- Looks like this will download gRPC without all the submodules, so I assume all the dependencies will be missing if you don't also do FetchContent_MakeAvailable() for gRPC's dependencies first?

Under the hood, does FetchContent_MakeAvailable basically do "add_subdirectory()" (=import grpc's CMakeLists.txt in the project's CMakeLists.txt) or does it basically build and install it into a scratch area?
The advantage of performing an actual installation under a prefix is that it's the "right C++ way" of using a dependency. If you just import the gRPC's build file in the top level build file, it can lead to some problems because the two CMakeLists.txt are not properly isolated (in which case I'm not sure if FetchContent is as great as it looks).

Reply:
In any case, once we agree whether FetchContent is what we recommend (currently I'm not 100% convinced), it would be good to add a few sentences that summarize what fetchContent does - this will be useful for folks that know how C++ dependencies work but are not that familiar with CMake.

Reply:
> download gRPC without all the submodules

FetchContent clones the submodules too.

> Under the hood, does FetchContent_MakeAvailable basically do "add_subdirectory()" (=import grpc's CMakeLists.txt in the project's CMakeLists.txt) or does it basically build and install it into a scratch area?

The former. In that regard it's very similar to setting up a git submodule for grpc.

> it would be good to add a few sentences that summarize what fetchContent does

👍

Reply:
Ok, so IIUC, FetchContent does:
1. download grpc and all its submodules
2. calls "cmake install" for gRPC's CMakeLists.txt (which is only useful with cmake 3.13+ because otherwise installation of gRPC's dependencies of from submodules doesn't work).
3. calls find_package(gRPC ....) to import installed gRPC's targets.

I assume you could also use FetchContent with submodules off and import all the gRPC's dependencies separately (e.g. FetchContent(c-ares), FetchContent(protobuf) etc...) and install all the deps one-by-one? 
The problem of installing everything as a monolith could be if user's project wanted also use some of gRPC's dependencies (in which case it's kind of annoying that gRPC's build and installation with gRPC_*_PROVIDER=module bundles all of them).

Overall I like the simplicity of FetchContent, but what I'm unsure about is whether offering too many different kind of installation options (which are kind of similar but not quite the same) will lead to confusion among users (and also on our team because it's hard to give advice when X different mode with subtle differences are supported).

IMHO the only really "clean" way of installing (from the purist perspective) is to install all the dependencies separately and then install gRPC (all under a prefix) and then use `find_package(gRPC ...)`, but it looks like `FetchContent` can't quite achieve that?
Superbuild can achieve that, but it's very annoying to use and uses weird kind of build layering. So I'm not sure what the approach here is.

Reply:
Ah, I misread the response. So it does add_subdirectory, which pretty much invalidates my previous comment.